modules/core/shared/src/main/scala/lucuma/core/data/Availability.scala

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2016-2025 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.core.data
+
+import cats.collections.Diet
+import cats.collections.Range
+import cats.kernel.Eq
+import cats.kernel.Semigroup
+import cats.syntax.all.*
+import lucuma.core.enums.Site
+import lucuma.core.instances.instant.given
+import lucuma.core.model.Semester
+import lucuma.core.util.DateInterval
+
+import java.time.Instant
+
+/**
+ * Wrapper for a `DisjointIntervalMap[Site, Instant]` with convenience methods for determining
+ * resource availability for a given site, time [range], or both.
+ */
+final case class Availability(toDisjointIntervalMap: DisjointIntervalMap[Site, Instant]):
+
+  /** The interval set for the given site. */
+  def forSite(site: Site): Diet[Instant] = 
+    toDisjointIntervalMap.get(site).getOrElse(Diet.empty)
+  
+  /** This `Availability`, narrowed to the specified range. */
+  def forRange(range: Range[Instant]): Availability =
+    Availability:
+      toDisjointIntervalMap.intersect(range)
+
+  /** This `Availability`, narrowed to the specified semester. */
+  def forSemester(semester: Semester): Availability =
+    Availability:
+      DisjointIntervalMap.unsafeFromMap:
+        Site.all.fproduct(forSiteAndSemester(_, semester)).toMap
+
+  /** This `Availability`, narrowed to the specified DateInterval. */
+  def forDateInterval(dateInterval: DateInterval): Availability =
+    Availability:
+      DisjointIntervalMap.unsafeFromMap:
+        Site.all.fproduct(forSiteAndDateInterval(_, dateInterval)).toMap  
+  
+  /** The interval set for the given site and DateInterval. */
+  def forSiteAndDateInterval(site: Site, dateInterval: DateInterval): Diet[Instant] =
+    val start = dateInterval.start.atStartOfDay(site.timezone).toInstant() 
+    val end   = dateInterval.end.atStartOfDay(site.timezone).plusDays(1).toInstant()
+    forSiteAndRange(site, Range(start, end))
```

Comment:
Since day boundaries occur in the middle of the night, I would if a `forObservingNightInterval` (`ObservingNight` includes a `Site`) would be useful at some point.